`unnecessary-lambda` false positive if parameter is used in lambda body

### Bug description

A `unnecessary-lambda` warning is created for lambda expressions which use one of its parameters in its body, for example:

```py
# pylint: disable=expression-not-assigned,missing-module-docstring,undefined-variable
lambda x: x(x)
```

My real code was something like

```py
lambda foo, bar: (function1 if bar == 1 else function2)(foo, bar)
```

### Configuration

_No response_

### Command used

```shell
pylint tmp.py
```


### Pylint output

```shell
************* Module tmp
tmp.py:2:0: W0108: Lambda may not be necessary (unnecessary-lambda)

------------------------------------------------------------------
Your code has been rated at 0.00/10 (previous run: 0.00/10, +0.00)
```


### Expected behavior

No warning

### Pylint version

```shell
pylint 2.17.0
astroid 2.15.0
Python 3.11.2 (main, Mar 17 2023, 12:30:14) [Clang 14.0.0 (clang-1400.0.29.202)]
```


### OS / Environment

_No response_

### Additional dependencies

_No response_

--- a/doc/whatsnew/fragments/8496.false_positive
+++ b/doc/whatsnew/fragments/8496.false_positive
@@ -1,0 +1,5 @@
+``unnecessary-lambda`` no longer warns on lambdas which use its parameters in
+their body (other than the final arguments), e.g.
+``lambda foo: (bar if foo else baz)(foo)``.
+
+Closes #8496

--- a/pylint/checkers/base/basic_checker.py
+++ b/pylint/checkers/base/basic_checker.py
@@ -576,6 +576,13 @@ class BasicChecker(_BasicChecker):
             if arg.name != passed_arg.name:
                 return
 
+        # The lambda is necessary if it uses its parameter in the function it is
+        # calling in the lambda's body
+        # e.g. lambda foo: (func1 if foo else func2)(foo)
+        for name in call.func.nodes_of_class(nodes.Name):
+            if name.lookup(name.name)[0] is node:
+                return
+
         self.add_message("unnecessary-lambda", line=node.fromlineno, node=node)
 
     @utils.only_required_for_messages("dangerous-default-value")

--- a/pylint/checkers/base/basic_checker.py
+++ b/pylint/checkers/base/basic_checker.py
@@ -519,6 +519,7 @@ class BasicChecker(_BasicChecker):
         )
 
     @utils.only_required_for_messages("unnecessary-lambda")
+    # pylint: disable-next=too-many-return-statements
     def visit_lambda(self, node: nodes.Lambda) -> None:
         """Check whether the lambda is suspicious."""
         # if the body of the lambda is a call expression with the same

--- a/tests/functional/u/unnecessary/unnecessary_lambda.py
+++ b/tests/functional/u/unnecessary/unnecessary_lambda.py
@@ -24,6 +24,12 @@ _ = lambda *args, **kwargs: _ANYARGS(*args, **kwargs)
 # +1: [unnecessary-lambda]
 _ = lambda x, y, z, *args, **kwargs: _ANYARGS(x, y, z, *args, **kwargs)
 
+# These don't use their parameters in their body
+# +1: [unnecessary-lambda]
+_ = lambda x: z(lambda x: x)(x)
+# +1: [unnecessary-lambda]
+_ = lambda x, y: z(lambda x, y: x + y)(x, y)
+
 # Lambdas that are *not* unnecessary and should *not* trigger warnings.
 _ = lambda x: x
 _ = lambda x: x()
@@ -50,3 +56,8 @@ _ = lambda: _ANYARGS(func=42)
 _ = lambda: code().analysis()
 
 _ = lambda **kwargs: dict(bar=42, **kwargs)
+
+# These use the lambda parameters in their body
+_ = lambda x: x(x)
+_ = lambda x, y: x(x, y)
+_ = lambda x: z(lambda y: x + y)(x)

--- a/tests/functional/u/unnecessary/unnecessary_lambda.txt
+++ b/tests/functional/u/unnecessary/unnecessary_lambda.txt
@@ -5,3 +5,5 @@ unnecessary-lambda:19:4:19:33:<lambda>:Lambda may not be necessary:UNDEFINED
 unnecessary-lambda:21:4:21:39:<lambda>:Lambda may not be necessary:UNDEFINED
 unnecessary-lambda:23:4:23:53:<lambda>:Lambda may not be necessary:UNDEFINED
 unnecessary-lambda:25:4:25:71:<lambda>:Lambda may not be necessary:UNDEFINED
+unnecessary-lambda:29:4:29:31:<lambda>:Lambda may not be necessary:UNDEFINED
+unnecessary-lambda:31:4:31:44:<lambda>:Lambda may not be necessary:UNDEFINED

